# Objects and operations

Python is a a popular scientific programming language and is regularly used both inside and outside of academia. 
In this course, we hope to introduce programming in Python before going on to show some applications of the langauge in a computational chemistry setting. 
The vision for the course structure is something a bit like that shown in {numref}`Figure {number} <content>`.
This means that we need to start by introducing some Python.

```{figure} ./images/content.png
---
scale: 35%
name: content
alt: A plot of the chemistry content increasing linearly over time while the programming content decreases at the same rate.
---
What you can expect to be learning over time.
```

Python consists of objects of *types*, where the type of an object may determine the operations that may be performed on it. 
For example, let us consider a simple numerical type, the integer.

In [1]:
4

4

The object in the above cell is an integer (or `int`), we can get the type of this object with the `type()` function.

In [2]:
type(4)

int

Since this object is numerical in nature, we can imagine using it in mathematical operations. 

In [3]:
4 + 4

8

In [4]:
type(4 + 4)

int

Above, we have used a mathematical operation between two integers to create a new integer object.
Integers are whole numbers, numbers that do not have a decimal point (e.g. 1, 2, 3, 4, 5, 6, etc.). 
However, Python is also capable of handling floating point (or `float`) type numbers, where there is a decimal point.

In [5]:
3.14

3.14

In [6]:
type(3.14)

float

These floating point numbers can also be used in mathematical operations.

In [7]:
1.1 + 2.2

3.3000000000000003

Notice, above the result of `1.1 + 2.2` is `3.3000000000000003`, which isn't exactly correct. 
This results from a problem of the way that floating point numbers are stored in the computer's memory, known as [floating point arthimetic](https://floating-point-gui.de). 
You may see examples of this throughout this course, typically the error in the result is **very** small. 
In the example above the result is off by less than $1\times 10^{-15}$ of a percent. 
Therefore, it is unlikely that this should impact the computation that you perform in this course. 
However, it is important to be aware of this problem.

> **Exercise** : In a Jupyter Notebook on your JupyterHub instance, consider the difference between `1`, `1.` and `1.0` as interpreted by Python.
> Use the `type()` function to determine the type of each of the values.
> From this, explain (in a Markdown cell in the same notebook) the importance of the decimal point in teh definition of numerical types in Python.

Textual information in Python is stored as in strings (or `str`), such as that shown below.

In [8]:
"Hello World!"

'Hello World!'

`"Hello World!"` is a [common first program](https://en.wikipedia.org/wiki/%22Hello,_World!%22_program) to write when learning a new programming language. 
> **Exercise**: Investigate the use of mathematical operations on strings. 
> Try adding multiple strings together and rationalise to yourself the result of this operation (searching online about "string concatention" might help this). 
> Also see if it is possible to subtract (`-`) one string from another and multiply (`*`) one string by an integer. 